src/main/com/yetanalytics/datasim/math/random.clj

```diff
@@ -1,211 +1,238 @@
 (ns com.yetanalytics.datasim.math.random
-  "Seeded random functions"
-  (:require [com.yetanalytics.datasim.util.maths :as maths]
-            [clojure.spec.alpha :as s]
-            [clojure.spec.gen.alpha :as sgen])
-  (:import [java.util
-            UUID Random]))
+  "Random number generation and probabilistic operations."
+  (:require [clojure.spec.alpha :as s]
+            [clojure.spec.gen.alpha :as sgen]
+            [com.yetanalytics.datasim.util.maths :as maths])
+  (:refer-clojure :exclude [rand rand-int rand-nth random-sample shuffle])
+  (:import [java.util UUID Random]))
+
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+;; Specs
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
 
 (s/def ::seed int?)
 
 (s/def ::rng
   (s/with-gen #(instance? Random %)
-    (fn []
-      (sgen/fmap (fn [s]
-                   (Random. s))
-                 (s/gen ::seed)))))
+    (fn [] (sgen/fmap (fn [s] (Random. s)) (s/gen ::seed)))))
+
+(s/def ::sd
+  (s/double-in :min 0.0 :infinite? false :NaN? false))
+
+(s/def ::prob
+  (s/double-in :min 0.0 :max 1.0))
+
+(s/def ::weight
+  (s/double-in :min -1.0 :max 1.0))
+
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+;; Functions
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+
+;; RNG Creation ;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+
+(s/fdef seed-rng
+  :args (s/cat)
+  :ret ::rng)
+
+(defn rng
+  "Create a pseudorandom RNG using an arbitrary seed value."
+  ^Random []
+  (Random.))
 
 (s/fdef seed-rng
   :args (s/cat :seed ::seed)
   :ret ::rng)
```

Comment:
Is there a duped fdef here?

Reply:
Not anymore: 07f303055955a6bc631b826c8acf8193a7cd1e0e